File: packages/react-native-nitro-vto/ios/NitroVtoView.swift
import UIKit
import ARKit
import MetalKit

/**
 * NitroVtoView - A UIView containing the AR glasses try-on view.
 *
 * This view handles:
 * - ARKit session management for face tracking
 * - Filament rendering via ArKitVtoAdapter
 * - Face tracking and glasses overlay
 *
 * Note: Camera permissions must be handled by the consuming React Native app
 * before this view becomes active.
 */
class NitroVtoView: UIView {

    private static let TAG = "NitroVtoView"

    // ARKit session
    private var arSession: ARSession?

    // Metal view for rendering
    private var metalView: MTKView?
    private var metalDevice: MTLDevice?

    // Filament renderer adapter
    private var arKitVtoAdapter: ArKitVtoAdapter?

    // Configuration
    private var modelUrl: String = ""
    private var isActiveState: Bool = true
    private var faceMeshOcclusionState: Bool = true
    private var backPlaneOcclusionState: Bool = true
    private var forwardOffsetState: Float = 0.005
    private var debugState: Bool = false

    // Callbacks
    var onModelLoaded: ((String) -> Void)?

    // State
    private var isInitialized = false
    private var isResumed = false

    // Display link for rendering
    private var displayLink: CADisplayLink?

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupMetalView()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupMetalView()
    }

    private func setupMetalView() {
        guard let device = MTLCreateSystemDefaultDevice() else {
            print("\(NitroVtoView.TAG): Metal is not supported on this device")
            return
        }
        metalDevice = device

        let mtkView = MTKView(frame: bounds, device: device)
        mtkView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mtkView.backgroundColor = .clear
        mtkView.framebufferOnly = false
        mtkView.isPaused = true
        mtkView.enableSetNeedsDisplay = false
        addSubview(mtkView)
        metalView = mtkView
    }

    // MARK: - Public API

    func setModelUrl(_ url: String) {
        if modelUrl != url {
            modelUrl = url
            if isInitialized {
                arKitVtoAdapter?.switchModel(withUrl: modelUrl)
            }
        }
    }

    func setIsActive(_ active: Bool) {
        if isActiveState != active {
            isActiveState = active
            if active && isResumed {
                resume()
            } else if !active {
                pause()
            }
        }
    }

    func switchModel(modelUrl: String) {
        self.modelUrl = modelUrl
        arKitVtoAdapter?.switchModel(withUrl: modelUrl)
    }

    func resetSession() {
        arKitVtoAdapter?.resetSession()
        if let session = arSession {
            let configuration = createARConfiguration()
            session.run(configuration, options: [.resetTracking, .removeExistingAnchors])
        }
    }

    func setFaceMeshOcclusion(_ enabled: Bool?) {
        faceMeshOcclusionState = enabled ?? true
        arKitVtoAdapter?.setFaceMeshOcclusion(faceMeshOcclusionState)
    }

    func setBackPlaneOcclusion(_ enabled: Bool?) {
        backPlaneOcclusionState = enabled ?? true
        arKitVtoAdapter?.setBackPlaneOcclusion(backPlaneOcclusionState)
    }

    func setForwardOffset(_ offset: Double?) {
        forwardOffsetState = Float(offset ?? 0.005)
        arKitVtoAdapter?.setForwardOffset(forwardOffsetState)
    }

    func setDebug(_ enabled: Bool?) {
        debugState = enabled ?? false
        arKitVtoAdapter?.setDebug(debugState)
    }

    // MARK: - Initialization

    private func initialize() {
        guard !isInitialized else { return }
        guard let mtkView = metalView else {
            print("\(NitroVtoView.TAG): Metal not available")
            return
        }

        // Create and initialize renderer
        arKitVtoAdapter = ArKitVtoAdapter(metalView: mtkView)
        arKitVtoAdapter?.onModelLoaded = onModelLoaded
        arKitVtoAdapter?.initialize(withModelUrl: modelUrl)

        // Apply stored configuration states
        arKitVtoAdapter?.setFaceMeshOcclusion(faceMeshOcclusionState)
        arKitVtoAdapter?.setBackPlaneOcclusion(backPlaneOcclusionState)
        arKitVtoAdapter?.setForwardOffset(forwardOffsetState)
        arKitVtoAdapter?.setDebug(debugState)

        isInitialized = true
        print("\(NitroVtoView.TAG): NitroVtoView initialized")
    }

    // MARK: - Lifecycle

    func resume() {
        isResumed = true

        guard isActiveState else { return }

        // Initialize if not already done
        if !isInitialized {
            initialize()
        }

        // Setup AR session if needed
        setupARSession()

        // Start display link for rendering
        startDisplayLink()

        // Resume renderer
        arKitVtoAdapter?.resume()
    }

    func pause() {
        stopDisplayLink()
        arKitVtoAdapter?.pause()
        arSession?.pause()
        isResumed = false
    }

    func destroy() {
        stopDisplayLink()
        arSession?.pause()
        arSession = nil
        arKitVtoAdapter?.destroy()
        arKitVtoAdapter = nil
        isInitialized = false
    }

    // MARK: - Display Link

    private func startDisplayLink() {
        guard displayLink == nil else { return }
        displayLink = CADisplayLink(target: self, selector: #selector(render))
        displayLink?.preferredFramesPerSecond = 60
        displayLink?.add(to: .main, forMode: .common)
    }

    private func stopDisplayLink() {
        displayLink?.invalidate()
        displayLink = nil
    }

    @objc private func render() {
        guard isInitialized, isActiveState else { return }
        guard let session = arSession, let frame = session.currentFrame else { return }

        // Get tracked faces
        let faces = frame.anchors.compactMap { $0 as? ARFaceAnchor }
            .filter { $0.isTracked }

        // Update renderer with current frame
        arKitVtoAdapter?.render(with: frame, faces: faces)
    }

    // MARK: - ARKit Setup

    private func setupARSession() {
        if let session = arSession {
            let configuration = createARConfiguration()
            session.run(configuration)
            return
        }

        guard ARFaceTrackingConfiguration.isSupported else {
            print("\(NitroVtoView.TAG): Face tracking is not supported on this device")
            return
        }

        let session = ARSession()
        arSession = session

        let configuration = createARConfiguration()
        session.run(configuration)

        print("\(NitroVtoView.TAG): ARKit session created successfully")
    }

    private func createARConfiguration() -> ARFaceTrackingConfiguration {
        let configuration = ARFaceTrackingConfiguration()
        configuration.isLightEstimationEnabled = true
        if #available(iOS 13.0, *) {
            configuration.maximumNumberOfTrackedFaces = 1
        }
        return configuration
    }

    // MARK: - View Lifecycle

    override func willMove(toWindow newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow)
        if newWindow == nil {
            // View is being removed
            destroy()
        }
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        metalView?.frame = bounds

        // Use drawable size (in pixels) not bounds (in points) for proper Retina support
        if let mtkView = metalView {
            let scale = mtkView.contentScaleFactor
            let widthPixels = Int32(bounds.width * scale)
            let heightPixels = Int32(bounds.height * scale)
            arKitVtoAdapter?.setViewportSizeWithWidth(widthPixels, height: heightPixels)
        }
    }
}

File: packages/react-native-nitro-vto/android/src/main/java/com/margelo/nitro/nitrovto/NitroVtoView.kt
package com.margelo.nitro.nitrovto

import android.content.Context
import android.util.Log
import android.view.Choreographer
import android.view.SurfaceView
import android.widget.FrameLayout
import com.google.ar.core.ArCoreApk
import com.google.ar.core.Config
import com.google.ar.core.Session
import com.google.ar.core.exceptions.CameraNotAvailableException
import com.google.ar.core.exceptions.UnavailableApkTooOldException
import com.google.ar.core.exceptions.UnavailableArcoreNotInstalledException
import com.google.ar.core.exceptions.UnavailableDeviceNotCompatibleException
import com.google.ar.core.exceptions.UnavailableSdkTooOldException
import java.util.EnumSet

/**
 * NitroVtoView - A FrameLayout containing the AR glasses try-on view.
 *
 * This view handles:
 * - ARCore session management
 * - Filament rendering via ArCoreVtoAdapter
 * - Face tracking and glasses overlay
 *
 * Note: Camera permissions must be handled by the consuming React Native app
 * before this view becomes active.
 */
class NitroVtoView(context: Context) : FrameLayout(context) {

    companion object {
        private const val TAG = "NitroVtoView"
    }

    // ARCore session
    private var arSession: Session? = null

    // SurfaceView for rendering
    private val surfaceView: SurfaceView = SurfaceView(context)

    // Filament renderer
    private var arCoreVtoAdapter: ArCoreVtoAdapter? = null

    // Configuration
    private var modelUrl: String = ""
    private var isActive: Boolean = true
    private var faceMeshOcclusionState: Boolean = true
    private var backPlaneOcclusionState: Boolean = true
    private var forwardOffsetState: Float = 0.005f
    private var debugState: Boolean = false

    // Callbacks
    var onModelLoaded: ((modelUrl: String) -> Unit)? = null

    // State
    private var isInitialized = false
    private var isResumed = false

    private val choreographer = Choreographer.getInstance()
    private val frameCallback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (!isResumed || !isActive || !isInitialized) {
                return
            }
            arCoreVtoAdapter?.renderOnce()
            choreographer.postFrameCallback(this)
        }
    }

    init {
        // Add SurfaceView to fill the entire view
        addView(surfaceView, LayoutParams(
            LayoutParams.MATCH_PARENT,
            LayoutParams.MATCH_PARENT
        ))
    }

    /**
     * Set the model URL for the glasses
     */
    fun setModelUrl(url: String) {
        if (modelUrl != url) {
            modelUrl = url
            if (isInitialized) {
                arCoreVtoAdapter?.switchModel(modelUrl)
            }
        }
    }

    /**
     * Set whether the AR session is active
     */
    fun setIsActive(active: Boolean) {
        if (isActive != active) {
            isActive = active
            if (active && isResumed) {
                resume()
            } else if (!active) {
                pause()
            }
        }
    }

    /**
     * Set face mesh occlusion enabled
     */
    fun setFaceMeshOcclusion(enabled: Boolean?) {
        faceMeshOcclusionState = enabled ?: true
        arCoreVtoAdapter?.setFaceMeshOcclusion(faceMeshOcclusionState)
    }

    /**
     * Set back plane occlusion enabled
     */
    fun setBackPlaneOcclusion(enabled: Boolean?) {
        backPlaneOcclusionState = enabled ?: true
        arCoreVtoAdapter?.setBackPlaneOcclusion(backPlaneOcclusionState)
    }

    /**
     * Set forward offset for glasses positioning (in meters)
     */
    fun setForwardOffset(offset: Double?) {
        forwardOffsetState = (offset ?: 0.005).toFloat()
        arCoreVtoAdapter?.setForwardOffset(forwardOffsetState)
    }

    /**
     * Set debug mode enabled
     */
    fun setDebug(enabled: Boolean?) {
        debugState = enabled ?: false
        arCoreVtoAdapter?.setDebug(debugState)
    }

    /**
     * Switch to a different glasses model
     */
    fun switchModel(modelUrl: String) {
        this.modelUrl = modelUrl
        arCoreVtoAdapter?.switchModel(modelUrl)
    }

    /**
     * Reset the AR session
     */
    fun resetSession() {
        arCoreVtoAdapter?.resetSession()
        arSession?.pause()
        arSession?.resume()
    }

    /**
     * Initialize the view. Should be called after the view is attached.
     */
    private fun initialize() {
        if (isInitialized) return

        // Create and initialize renderer
        arCoreVtoAdapter = ArCoreVtoAdapter(context)
        arCoreVtoAdapter?.onModelLoaded = onModelLoaded
        arCoreVtoAdapter?.initialize(surfaceView, modelUrl)
        arCoreVtoAdapter?.setFaceMeshOcclusion(faceMeshOcclusionState)
        arCoreVtoAdapter?.setBackPlaneOcclusion(backPlaneOcclusionState)
        arCoreVtoAdapter?.setForwardOffset(forwardOffsetState)
        arCoreVtoAdapter?.setDebug(debugState)

        isInitialized = true
        Log.d(TAG, "NitroVtoView initialized")
    }

    /**
     * Resume the AR session and rendering
     */
    fun resume() {
        isResumed = true

        if (!isActive) return

        // Initialize if not already done
        if (!isInitialized) {
            initialize()
        }

        // Setup AR session if needed
        setupArSession()

        // Resume renderer
        arCoreVtoAdapter?.resume()
        startFrameLoop()
    }

    /**
     * Pause the AR session and rendering
     */
    fun pause() {
        stopFrameLoop()
        arCoreVtoAdapter?.pause()
        arSession?.pause()
        isResumed = false
    }

    /**
     * Destroy and clean up resources
     */
    fun destroy() {
        stopFrameLoop()
        arSession?.close()
        arSession = null
        arCoreVtoAdapter?.destroy()
        arCoreVtoAdapter = null
        isInitialized = false
    }

    /**
     * Sets up the ARCore session with face tracking.
     * Assumes camera permission is already granted.
     */
    private fun setupArSession() {
        if (arSession != null) {
            arSession?.resume()
            arCoreVtoAdapter?.session = arSession
            return
        }

        try {
            // Check ARCore availability
            when (ArCoreApk.getInstance().requestInstall(getActivity(), true)) {
                ArCoreApk.InstallStatus.INSTALL_REQUESTED -> return
                ArCoreApk.InstallStatus.INSTALLED -> { /* Continue */ }
            }

            // Create AR session with front camera + shared camera stream support.
            // Fallback to plain front camera if shared camera is unavailable on this device/ARCore build.
            arSession = try {
                Session(context, EnumSet.of(Session.Feature.FRONT_CAMERA, Session.Feature.SHARED_CAMERA))
            } catch (error: Throwable) {
                Log.w(TAG, "Shared camera feature unavailable, falling back to FRONT_CAMERA only: ${error.message}")
                Session(context, EnumSet.of(Session.Feature.FRONT_CAMERA))
            }

            // Configure session for face tracking
            val config = Config(arSession).apply {
                augmentedFaceMode = Config.AugmentedFaceMode.MESH3D
                planeFindingMode = Config.PlaneFindingMode.DISABLED
                lightEstimationMode = Config.LightEstimationMode.AMBIENT_INTENSITY
                // Enable depth if supported by device
                depthMode = if (arSession!!.isDepthModeSupported(Config.DepthMode.AUTOMATIC)) {
                    Config.DepthMode.AUTOMATIC
                } else {
                    Config.DepthMode.DISABLED
                }
            }
            arSession?.configure(config)

            // Resume session
            arSession?.resume()

            // Connect session to renderer
            arCoreVtoAdapter?.session = arSession

            Log.d(TAG, "ARCore session created successfully")

        } catch (e: UnavailableArcoreNotInstalledException) {
            Log.e(TAG, "ARCore is not installed")
        } catch (e: UnavailableDeviceNotCompatibleException) {
            Log.e(TAG, "This device does not support AR")
        } catch (e: UnavailableSdkTooOldException) {
            Log.e(TAG, "Please update ARCore")
        } catch (e: UnavailableApkTooOldException) {
            Log.e(TAG, "Please update this app")
        } catch (e: CameraNotAvailableException) {
            Log.e(TAG, "Camera not available")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create AR session: ${e.message}")
        }
    }

    private fun startFrameLoop() {
        choreographer.removeFrameCallback(frameCallback)
        choreographer.postFrameCallback(frameCallback)
    }

    private fun stopFrameLoop() {
        choreographer.removeFrameCallback(frameCallback)
    }

    /**
     * Helper to get the activity from context
     */
    private fun getActivity(): android.app.Activity? {
        var ctx = context
        while (ctx is android.content.ContextWrapper) {
            if (ctx is android.app.Activity) {
                return ctx
            }
            ctx = ctx.baseContext
        }
        return null
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        Log.d(TAG, "onAttachedToWindow")
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        Log.d(TAG, "onDetachedFromWindow")
        destroy()
    }
}
